### PR TITLE
warn on unused withl clause

### PR DESCRIPTION
**Pull request by mat-hek**

### CODE REVIEW

File: lib/bunch.ex [hunk 1]
```diff
@@ -236,23 +237,38 @@ defmodule Bunch do
...
+             unquote(acc)
+           end, used_labels}
+      end)
+
+    unused_else_clauses = Map.drop(else_clauses, used_labels)
+
+    Enum.each(unused_else_clauses, fn {label, clauses} ->
+      Enum.each(clauses, fn {:->, meta, _args} ->
+        log_compile_warning(
+          "withl else clause labelled #{inspect(label)} will never match",
```
Comment (bblaszkow06) [inline lib/bunch.ex:264]:
```suggestion
          "withl's else clause labelled #{inspect(label)} will never match",
```

File: lib/bunch.ex [hunk 2]
```diff
@@ -236,23 +237,38 @@ defmodule Bunch do
...
-          unquote(acc)
-        end
+    {result, used_labels} =
+      with_clauses
+      |> Enum.reverse()
+      |> Enum.reduce({block, []}, fn
+        {label, {:<-, meta, _args} = clause}, {acc, used_labels} ->
+          label_else_clauses =
+            else_clauses[label] ||
+              "Label `#{inspect(label)}` not present in withl else clauses"
```
Comment (bblaszkow06) [inline lib/bunch.ex:247]: Since this is printed to the console, backticks seem redundant and you don't use them in line 264.

File: lib/bunch.ex [hunk 3]
```diff
@@ -236,23 +237,38 @@ defmodule Bunch do
...
+             unquote(clause)
+             unquote(acc)
+           end, used_labels}
+      end)
+
+    unused_else_clauses = Map.drop(else_clauses, used_labels)
+
+    Enum.each(unused_else_clauses, fn {label, clauses} ->
+      Enum.each(clauses, fn {:->, meta, _args} ->
+        log_compile_warning(
```
Comment (bblaszkow06) [inline lib/bunch.ex:263]: It would be nice to have some tests for it as it would also help me reproduce the warnings

File: lib/bunch.ex (hunk 2 above)
Comment (mat-hek) [reply lib/bunch.ex:247]:
```suggestion
              "Label #{inspect(label)} not present in withl else clauses"
```

File: lib/bunch.ex [hunk 4]
```diff
@@ -136,7 +136,27 @@ defmodule Bunch do
...
+      ...>     :ok
+      ...>   else
+      ...>     a: _a -> :error
+      ...>     b: _b -> :error
+      ...>     c: _c -> :error
+      ...>   end
+      ...> end) end)
+      iex> false = String.contains?(warns, "withl's else clause labelled :a will never match")
+      iex> true = String.contains?(warns, "withl's else clause labelled :b will never match")
+      iex> true = String.contains?(warns, "withl's else clause labelled :c will never match")
```
Comment (bblaszkow06) [inline lib/bunch.ex:159]: It looks awful in docs, you should rather put it in test

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

**Review by bblaszkow06 — APPROVED** [review]

**Pull request merged**